src/MOI_wrapper.jl

```diff
@@ -118,397 +210,1291 @@ end
+MOI.supports(::Optimizer, ::MOI.ConstraintDual, c) = true
+MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
+MOI.supports(::Optimizer, ::MOI.ListOfConstraintIndices) = true
+MOI.supports(::Optimizer, ::MOI.RawStatusString) = true
+
+MOI.Utilities.supports_default_copy_to(::Optimizer, ::Bool) = true
+
+function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike; kwargs...)
+    return MOI.Utilities.automatic_copy_to(dest, src; kwargs...)
+end
+
+function MOI.get(model::Optimizer, ::MOI.ListOfVariableAttributesSet)
+    return MOI.AbstractVariableAttribute[MOI.VariableName()]
+end
+
+function MOI.get(model::Optimizer, ::MOI.ListOfModelAttributesSet)
+    attributes = [
+        MOI.ObjectiveSense(),
+        MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}()
+    ]
+    if MOI.get(model, MOI.Name()) != ""
+        push!(attributes, MOI.Name())
     end
-    model.needs_update = false
-    model.num_variables = 0
-    model.num_linear_constraints = 0
-    model.num_quadratic_constraints = 0
-    return
-end
-
-LQOI.supported_constraints(::Optimizer) = SUPPORTED_CONSTRAINTS
-LQOI.supported_objectives(::Optimizer)  = SUPPORTED_OBJECTIVES
-
-LQOI.backend_type(::Optimizer, ::MOI.EqualTo{Float64})     = Cchar('=')
-LQOI.backend_type(::Optimizer, ::MOI.LessThan{Float64})    = Cchar('<')
-LQOI.backend_type(::Optimizer, ::MOI.GreaterThan{Float64}) = Cchar('>')
-LQOI.backend_type(::Optimizer, ::MOI.Zeros)                = Cchar('=')
-LQOI.backend_type(::Optimizer, ::MOI.Nonpositives)         = Cchar('<')
-LQOI.backend_type(::Optimizer, ::MOI.Nonnegatives)         = Cchar('>')
-
-function LQOI.change_variable_bounds!(model::Optimizer,
-          columns::Vector{Int}, new_bounds::Vector{Float64},
-          senses::Vector{Cchar})
-    number_lower_bounds = count(x->x==Cchar('L'), senses)
-    lower_cols   = fill(0, number_lower_bounds)
-    lower_values = fill(0.0, number_lower_bounds)
-    number_upper_bounds = count(x->x==Cchar('U'), senses)
-    upper_cols   = fill(0, number_upper_bounds)
-    upper_values = fill(0.0, number_upper_bounds)
-    lower_index = 1
-    upper_index = 1
-    for (column, bound, sense) in zip(columns, new_bounds, senses)
-        if sense == Cchar('L')
-            lower_cols[lower_index]   = column
-            lower_values[lower_index] = bound
-            lower_index += 1
-        elseif sense == Cchar('U')
-            upper_cols[upper_index]   = column
-            upper_values[upper_index] = bound
-            upper_index += 1
+    return attributes
+end
+
+function MOI.get(model::Optimizer, ::MOI.ListOfConstraintAttributesSet)
+    return MOI.AbstractConstraintAttribute[MOI.ConstraintName()]
+end
+
+function indices_and_coefficients(model::Optimizer, f::MOI.ScalarAffineFunction{Float64})
+    f_canon = MOI.Utilities.canonical(f)
+    indices = Int[]
+    coefficients = Float64[]
+    for term in f_canon.terms
+        push!(indices, model[term.variable_index].column)
```

Comment:
If no variable has been deleted, `model[term.variable_index].column` should be `term.variable_index.value` ? Needing a dictionary lookup for each term might be very slow
Related to https://github.com/JuliaOpt/LinQuadOptInterface.jl/issues/97

Reply:
If `MOI.empty!` has been called, this this is also needed. Having the indices keep growing helped find that bug in MOI.Test (https://github.com/JuliaOpt/MathOptInterface.jl/pull/763).

Reply:
That's the purpose of xor'ing indices in MockOptimizer, we should have run the copy tests with it